src/commands/apphosting-rollouts-create.ts

```diff
@@ -1,27 +1,116 @@
+    const commit = options.commit as string | undefined;
+    if (branch && commit) {
+      throw new FirebaseError(
+        "Cannot specify both a branch and commit to deploy. Please specify either --git-branch or --commit.",
+      );
+    }
+    const backend = await apphosting.getBackend(projectId, location, backendId);
+    if (!backend.codebase.repository) {
+      throw new FirebaseError(
+        `Failed to find connected Git repository for backend ${backendId}. Please check that your backend has been successfully linked to a Git repository.`,
+      );
+    }
+    const { repoLink, owner, repo, readToken } = await getRepoDetailsFromBackend(
+      projectId,
+      location,
+      backend.codebase.repository,
+    );
+
+    let target;
+    if (branch) {
+      const branches = await listAllBranches(repoLink.name);
+      if (!branches.has(branch)) {
+        throw new FirebaseError(
+          `Unrecognized git branch ${branch}. Please double-check your branch name and try again.`,
+        );
+      }
+      const branchInfo = await getGitHubBranch(owner, repo, branch, readToken.token);
+      target = branchInfo.commit;
+    } else if (commit) {
+      if (!/^(?:[0-9a-f]{40}|[0-9a-f]{7})$/.test(commit)) {
+        throw new FirebaseError(`Invalid git commit ${commit}. Must be a valid SHA1 hash.`);
+      }
+      try {
+        const commitInfo = await getGitHubCommit(owner, repo, commit, readToken.token);
+        target = commitInfo;
+      } catch (err: unknown) {
+        if ((err as FirebaseError).status === 422) {
+          throw new FirebaseError(
+            `Unrecognized git commit ${commit}. Please double-check your commit hash and try again.`,
+          );
+        }
+        throw err;
+      }
+    } else {
+      branch = await promptGitHubBranch(repoLink);
+      const branchInfo = await getGitHubBranch(owner, repo, branch, readToken.token);
+      target = branchInfo.commit;
+    }
+
+    logBullet(`You are about to deploy [${target.sha.substring(0, 7)}]: ${target.commit.message}`);
+    const confirmRollout = await confirm({
+      force: options.force,
+      message: "Do you want to continue?",
     });
-    logger.info(`Started a rollout for backend ${backendId} with build ${buildId}.`);
-    logger.info("Check status by running:");
-    logger.info(`\tfirebase apphosting:rollouts:list --location ${location}`);
-    return op;
+    if (!confirmRollout) {
+      return;
+    }
+    logBullet(
+      `You may also track this rollout at:\n\t${consoleOrigin()}/project/${projectId}/apphosting`,
+    );
+
+    const createRolloutSpinner = ora(
+      "Starting a new rollout; this may take a few minutes. It's safe to exit now.",
+    ).start();
+
+    let rollout;
+    try {
+      ({ rollout } = await orchestrateRollout({
+        projectId,
+        location,
+        backendId,
+        buildInput: {
+          source: {
+            codebase: {
+              commit: target.sha,
```

Comment:
this is making me realize we need an explicit type for the `target` declaration

Reply:
`target` is of type `GitHubCommitInfo`, which is defined in `src/apphosting/githubConnections.ts`. Exported the interface and explicitly typed `targetCommit: GitHubCommitInfo`.